src/dragonball/src/dbs_pci/src/configuration.rs

```diff
@@ -0,0 +1,1982 @@
+    Test = 0xFF,
+}
+
+impl From<u8> for PciCapabilityID {
+    fn from(c: u8) -> Self {
+        match c {
+            0 => PciCapabilityID::ListID,
+            0x01 => PciCapabilityID::PowerManagement,
+            0x02 => PciCapabilityID::AcceleratedGraphicsPort,
+            0x03 => PciCapabilityID::VitalProductData,
+            0x04 => PciCapabilityID::SlotIdentification,
+            0x05 => PciCapabilityID::MessageSignalledInterrupts,
+            0x06 => PciCapabilityID::CompactPCIHotSwap,
+            0x07 => PciCapabilityID::PCIX,
+            0x08 => PciCapabilityID::HyperTransport,
+            0x09 => PciCapabilityID::VendorSpecific,
+            0x0A => PciCapabilityID::Debugport,
+            0x0B => PciCapabilityID::CompactPCICentralResourceControl,
+            0x0C => PciCapabilityID::PCIStandardHotPlugController,
+            0x0D => PciCapabilityID::BridgeSubsystemVendorDeviceID,
+            0x0E => PciCapabilityID::AGPTargetPCIPCIbridge,
+            0x0F => PciCapabilityID::SecureDevice,
+            0x10 => PciCapabilityID::PCIExpress,
+            0x11 => PciCapabilityID::MSIX,
+            0x12 => PciCapabilityID::SATADataIndexConf,
+            0x13 => PciCapabilityID::PCIAdvancedFeatures,
+            0x14 => PciCapabilityID::PCIEnhancedAllocation,
+            #[cfg(test)]
+            0xFF => PciCapabilityID::Test,
+            _ => PciCapabilityID::ListID,
+        }
+    }
+}
+
+/// A PCI capability list.
+///
+/// Devices can optionally specify capabilities in their configuration space.
+#[allow(clippy::len_without_is_empty)]
+pub trait PciCapability: Send + Sync + Downcast {
+    /// Get size of the whole capability structure, including the capability header.
+    fn len(&self) -> usize;
+
+    /// Link to next PCI capability.
+    fn set_next_cap(&mut self, next: u8);
+
+    /// Read a 8bit value from the capability.
+    fn read_u8(&mut self, offset: usize) -> u8;
+
+    /// Read a 16bit value from the capability.
+    fn read_u16(&mut self, offset: usize) -> u16 {
+        self.read_u8(offset) as u16 | (self.read_u8(offset + 1) as u16) << 8
+    }
+
+    /// Read a 32bit value from the capability.
+    fn read_u32(&mut self, offset: usize) -> u32 {
+        self.read_u16(offset) as u32 | (self.read_u16(offset + 2) as u32) << 16
+    }
+
+    /// Write a 8bit value to the capability.
+    fn write_u8(&mut self, offset: usize, value: u8);
+
+    /// Write a 16bit value to the capability.
+    fn write_u16(&mut self, offset: usize, value: u16) {
+        self.write_u8(offset, value as u8);
+        self.write_u8(offset + 1, (value >> 8) as u8);
+    }
+
+    /// Write a 32bit value to the capability.
+    fn write_u32(&mut self, offset: usize, value: u32) {
+        self.write_u16(offset, value as u16);
+        self.write_u16(offset + 2, (value >> 16) as u16);
+    }
+
+    /// The type of PCI Interrupt
+    fn pci_capability_type(&self) -> PciCapabilityID;
+}
+
+downcast_rs::impl_downcast!(PciCapability);
```

Comment:
Do we need this?

Reply:
nope, but this will be used in later PRs and it's basic functionality for PciCapabilities.